libvast/builtins/stores/feather.cpp

```diff
@@ -92,60 +96,89 @@ auto wrap_record_batch(const table_slice& slice)
-    const auto read_status = reader.MoveValueUnsafe()->Read(&table);
-    if (!read_status.ok())
-      return caf::make_error(ec::system_error, read_status.ToString());
-
-    table_ = std::move(table);
-    auto arrow_field = table_->schema()->GetFieldByName("event");
-    if (!arrow_field) {
-      return caf::make_error(ec::format_error, "schema does not have mandatory "
-                                               "'event' column");
-    }
-    schema_ = type::from_arrow(*arrow_field);
-    if (!schema_)
-      return caf::make_error(ec::format_error,
-                             "Arrow schema incompatible with VAST type: {}",
-                             arrow_field->ToString(true));
-    num_events_ = table_->num_rows();
+    // See arrow::ipc::internal::kArrowMagicBytes in
+    // arrow/ipc/metadata_internal.h.
+    static constexpr auto arrow_magic_bytes = std::string_view{"ARROW1"};
+    if (chunk->size() < arrow_magic_bytes.length()
+        || std::memcmp(chunk->data(), arrow_magic_bytes.data(),
+                       arrow_magic_bytes.size())
+             != 0)
+      return caf::make_error(ec::format_error, "not an Apache Feather v1 or "
+                                               "Arrow IPC file");
+    remaining_slices_generator_ = decode_ipc_stream(std::move(chunk));
+    remaining_slices_iterator_ = remaining_slices_generator_.begin();
     return {};
   }
 
   [[nodiscard]] detail::generator<table_slice> slices() const override {
     auto offset = id{};
-    for (auto rb : arrow::TableBatchReader(*table_)) {
-      VAST_ASSERT(rb.ok(), rb.status().ToString().c_str());
-      auto time_col = rb.ValueUnsafe()->GetColumnByName("import_time");
-      auto unwrapped_rb = unwrap_record_batch(rb.MoveValueUnsafe());
-      auto slice = table_slice{unwrapped_rb, schema_};
-      slice.offset(offset);
-      offset += slice.rows();
-      slice.import_time(derive_import_time(time_col));
-      co_yield std::move(slice);
+    auto i = size_t{};
+    while (true) {
+      if (i < cached_slices_.size()) {
+        VAST_ASSERT(offset == cached_slices_[i].offset());
+      } else if (remaining_slices_iterator_
+                 != remaining_slices_generator_.end()) {
+        auto batch = std::move(*remaining_slices_iterator_);
+        VAST_ASSERT(batch);
+        ++remaining_slices_iterator_;
+        auto import_time_column = batch->GetColumnByName("import_time");
+        auto slice = cached_slices_.empty()
+                       ? table_slice{unwrap_record_batch(batch)}
+                       : table_slice{unwrap_record_batch(batch),
+                                     cached_slices_[0].layout()};
+        slice.offset(offset);
+        slice.import_time(derive_import_time(import_time_column));
+        cached_slices_.push_back(std::move(slice));
+      } else {
+        co_return;
+      }
+      co_yield cached_slices_[i];
+      offset += cached_slices_[i].rows();
+      ++i;
```

Comment:
Meta-comment: If this is now much faster than before than we should write a changelog entry.

Reply:
When monitoring the `passive-store.init.runtime` metric, it is now consistently <10ms per store when previously during peak load it averaged ~500ms per store for me when testing on a database with 300M events in a test that rebuilds the whole database (i.e., a stress test).

I then ran `vast export -n 1 json` 10 times for a cold start of VAST with both master and this PR 20 times each on the same database, and the averaged results positively surprised me:

||master|this PR|
|-:|-:|-:|
|stores loaded|10|5|
|per store init time|55.1ms|4.2ms|
|total store init time|551.2ms|21.0ms|
|time to first result|19.8ms|14.5ms|

I would say that these results are significant. It doesn't make VAST much faster to the user because we still have the signal monitor bug in VAST that Tobias wanted to fix, but regardless of that it makes queries that are done quickly or cancelled now take significantly less CPU time in VAST, and in general makes loading a store much more friendly to the actor scheduler, which _should_ make the whole system feel a bit more responsive under high load.

Reply:
Some further testing using the command `vast export null`, again on a cold start of VAST averaged over 20 runs:

||master|this PR|
|-:|-:|-:|
|time to first result|69.2ms|25.4ms|
|total time|39.38s|33.30s|
|per store init time|386.5ms|7.3ms|

Yeah, this definitely needs a changelog entry.

Reply:
> To summarize: Yeah, this definitely needs a changelog entry.

I would put this comment directly in the next release blog post, even within this PR.

Reply:
I added a changelog entry, a blog post for v2.4.1, and bumped versions.

All this needs is approval and it needs to be cherry-picked onto a v2.4.x branch, and then we can cut a new release tomorrow that includes just this over v2.4.0.